lib/Sema/CSBindings.cpp

```diff
@@ -25,22 +25,26 @@ Optional<ConstraintSystem::PotentialBindings>
 ConstraintSystem::determineBestBindings() {
   // Look for potential type variable bindings.
   Optional<PotentialBindings> bestBindings;
-  llvm::SmallDenseMap<TypeVariableType *, PotentialBindings> cache;
-
   // First, let's collect all of the possible bindings.
   for (auto *typeVar : getTypeVariables()) {
     if (typeVar->getImpl().hasRepresentativeOrFixed())
       continue;
 
+    if (Bindings.count(typeVar) > 0)
+      continue;
+
     if (auto bindings = getPotentialBindings(typeVar))
-      cache.insert({typeVar, std::move(bindings)});
+      Bindings.insert({typeVar, std::move(bindings)});
   }
 
   // Now let's see if we could infer something for related type
   // variables based on other bindings.
   for (auto *typeVar : getTypeVariables()) {
-    auto cachedBindings = cache.find(typeVar);
-    if (cachedBindings == cache.end())
+    if (typeVar->getImpl().hasRepresentativeOrFixed())
+      continue;
+
+    auto cachedBindings = Bindings.find(typeVar);
```

Comment:
auto &?

Reply:
Ah yes, thank you! That’s an existing problem though

Reply:
Actually reference can't be used here because `find` return an iterator...